llo/plugin.go

```diff
@@ -0,0 +1,924 @@
+// observations.
+//
+// This function should be pure. Don't do anything slow in here.
+//
+// You may assume that the outctx.SeqNr is increasing monotonically (though
+// *not* strictly) across the lifetime of a protocol instance and that
+// outctx.previousOutcome contains the consensus outcome with sequence
+// number (outctx.SeqNr-1).
+func (p *LLOPlugin) Outcome(outctx ocr3types.OutcomeContext, query types.Query, aos []types.AttributedObservation) (ocr3types.Outcome, error) {
+	if len(aos) == 0 {
+		return nil, errors.New("no attributed observations")
+	}
+
+	if outctx.SeqNr <= 1 {
+		// Initial Outcome
+		var lifeCycleStage commontypes.LLOLifeCycleStage
+		if p.PredecessorConfigDigest == nil {
+			// Start straight in production if we have no predecessor
+			lifeCycleStage = LifeCycleStageProduction
+		} else {
+			lifeCycleStage = LifeCycleStageStaging
+		}
+		outcome := Outcome{
+			lifeCycleStage,
+			0,
+			nil,
+			nil,
+			nil,
+		}
+		return json.Marshal(outcome)
+	}
+
+	/////////////////////////////////
+	// Decode previousOutcome
+	/////////////////////////////////
+	var previousOutcome Outcome
+	if err := json.Unmarshal(outctx.PreviousOutcome, &previousOutcome); err != nil {
+		return nil, fmt.Errorf("error unmarshalling previous outcome: %v", err)
+	}
+
+	/////////////////////////////////
+	// Decode observations
+	/////////////////////////////////
+
+	// a single valid retirement report is enough
+	var validPredecessorRetirementReport *RetirementReport
+
+	shouldRetireVotes := 0
+
+	timestampsNanoseconds := []int64{}
+
+	removeChannelVotesByID := map[commontypes.ChannelID]int{}
+
+	// for each channelId count number of votes that mention it and count number of votes that include it.
+	addChannelVotesByHash := map[ChannelHash]int{}
+	addChannelDefinitionsByHash := map[ChannelHash]ChannelDefinitionWithID{}
+
+	streamObservations := map[commontypes.StreamID][]*big.Int{}
+
+	for _, ao := range aos {
+		observation := Observation{}
+		// TODO: Use protobufs
+		if err := json.Unmarshal(ao.Observation, &observation); err != nil {
+			p.Logger.Warnw("ignoring invalid observation", "oracleID", ao.Observer, "error", err)
+			continue
+		}
+
+		if len(observation.AttestedPredecessorRetirement) != 0 && validPredecessorRetirementReport == nil {
+			pcd := *p.PredecessorConfigDigest
+			retirementReport, err := p.PredecessorRetirementReportCache.CheckAttestedRetirementReport(pcd, observation.AttestedPredecessorRetirement)
+			if err != nil {
+				p.Logger.Warnw("ignoring observation with invalid attested predecessor retirement", "oracleID", ao.Observer, "error", err, "predecessorConfigDigest", pcd)
+				continue
+			}
+			validPredecessorRetirementReport = &retirementReport
+		}
+
+		if observation.ShouldRetire {
+			shouldRetireVotes++
+		}
+
+		timestampsNanoseconds = append(timestampsNanoseconds, observation.UnixTimestampNanoseconds)
+
+		for channelID := range observation.RemoveChannelIDs {
+			removeChannelVotesByID[channelID]++
```

Comment:
should we verify the channel-id exists in `removeChannelVotesByID` - idk if it's possible for a faulty/malicious observation to come in and break this flow by adding a non-existent channel id

Reply:
We don't need to verify it exists, in fact this map is empty so none of them will exist.

We need to handle the scenario that the node executing this code is:
1) Honest
2) May have an outdated/incorrect view of channel definitions

We come to consensus on which channels should be added/removed. It requires that at least 2f+1 nodes are honest and have a proper view on channel definitions. So even if you have a malicious node put whatever garbage it wants here, unless 2f+1 are in agreement, it won't be applied.